accepted/async-imports.md

```diff
@@ -0,0 +1,421 @@
+* [Lunatic](https://github.com/lunatic-lang/lunatic) is an example project using
+  Wasmtime which has stack switching under the hood.
+
+A frequent use case of Wasmtime is to power server-side WebAssembly execution
+where the server may be executing many WebAssembly modules simultaneously. It's
+expected that many of these modules will be waiting on I/O or host
+functionality, and having a thread-per-module will be too costly in terms of
+synchronization and resources.
+
+The goal of this RFC is to specifically target the `async` use case in Rust.
+While the implementation will involve stack-switching and some communication
+between the two stacks, it does not expose an API to users which allows
+manually doing low-level operations like stack switching, yielding, and
+resumption.
+
+# Proposal
+[proposal]: #proposal
+
+This RFC proposes adding `async` support natively to the `wasmtime` crate
+itself. While it's perhaps possible to build this externally (as shown by
+`wasmtime-async`) it's a requested enough feature that adding core support
+should increase its visibility and help it feel more well supported.
+
+When adding support for the `wasmtime` crate, however, this proposal also aims
+to not disturb the functionality that Wasmtime has today. Not all users of
+Wasmtime need to handle `async` code and/or switch stacks. It should still be
+possible, for example if necessary, to build Wasmtime without `async` support.
+This effectively means that this proposal isn't changing existing APIs in
+`wasmtime`, instead augmenting existing ones. Additionally, however, we want to
+avoid "split the world" scenarios as much as we can. For example `Memory`
+doesn't have much reason to be different in async and non-async worlds, and same
+with `Module`.
+
+### Changes to `Store`
+
+With these goals in mind, the first change to Wasmtime is a new
+constructor for `Store`:
+
+```rust
+impl Store {
+    pub fn new_async(engine: &Engine) -> Store {
+        // ...
+    }
+}
+```
+
+This `new_async` method sets a runtime flag within `Store` intending to indicate
+that it is to be used with asynchronous methods defined below. The goal of this
+method is to prevent mistakes, not to enable any sort of internal functionality
+at this time. This'll be a bit more clear below as we cover the changes to
+`Func`.
+
+### Creating a `Func`
+
+The first location where dealing with `async` functions on the host comes into
+play when you're creating a `Func` from host-defined functionality. The existing
+`Func::new` and `Func::wrap` constructors will continue to work in both async
+and non-async `Store` objects. In additon to these constructors, however, new
+constructors which take asynchronous functions will be added:
+
+```rust
+impl Func {
+    pub fn new_async<'a, T, R>(
+        store: &Store,
+        ty: FuncType,
+        state: T,
+        func: fn(Caller<'a>, &'a T, &'a [Val], &'a mut [Val]) -> R,
+    ) -> Func
```

Comment:
Should this (and `wrap_async`) be `-> Result<Func>` so that it can gracefully fail when given a non-async `Store`?

Reply:
Erm, should have kept reading to where you talk about panicking. Not sure I like the idea of panicking for the mismatch, but I can see the logic in doing so.

Reply:
It's true yeah that either way is technically possible, but I would personally tend to classify this as a programmer error because applications would be expected to pick one or the other of async or not, and if you get it wrong by accident it seems like you'd want to discover that ASAP

Reply:
I'm onboard with panicking 👍

Reply:
I agree with panicking on mismatch, if async is an internal implementation detail of the `Store`.

Is there any way, though, that we could capture this in the type system, and then produce a type error at compile time?

I can think of two ways to do that. (All names below are drafts, intended as placeholders; we can always change them.)

- We could make `Store` take a type parameter for a marker type, where most functions accept `Store<A>` for any `A`, but functions that require an async-capable store *require* `Store<AsyncStore>`. Most `Store` functions would be implemented for `Store<A>` for any `A`, but `Store::new` and `Store::new_async` would return one of the two specific types. Then, if you attempt to call `Func::new_async` with a `Store<SyncStore>` you'll get a compile error. This has two advantages: `Store` is the same type with no duplicate implementation effort, and the async property becomes a compile-time property rather than a run-time boolean. Most code will only ever use one of those two types, so this should not add to code size; any given program should only monomorphize one of those two types. The disadvantage would be having to pass around the generic parameter, but that seems worth getting this checked at compile time. We could also have type aliases for the two types.

- We could have two types, `Store` and `StoreAsync`, and implement a sealed trait `AnyStore` on them (so that those are the only two instances). Most functions would accept `impl AnyStore`, and functions like `Func::new_async` would accept only `StoreAsync` specifically. This has the advantage of not requiring a generic parameter everywhere. The disadvantage is that we'd need a more complex set of forwarding functions within wasmtime to handle the two distinct types, so while we could share one implementation, we'll have 2-3 sets of function signatures and simple forwarding stubs for many of the same functions.

Personally, I'd favor the former approach: `Store<SyncStore>`, `Store<AsyncStore>`, and type aliases for those two so people don't normally have to think about the generic parameter.

Thoughts?

Reply:
I think it's possible to move this into the type system but I'd fear that we're kind of shooting ourselves in the foot too much and/or sacrificing ergonomics. We would not only need an async store type-level distinction but we would also need and async function type-level distinction because if you extract a function from an instance then you can only call it asynchronously in an async store. But if we don't want a runtime footgun then we'd also need async instances, and this quickly spirals to creating almost double the api surface area.

Overall it seemed that given the use cases in mind a runtime flag would be fine here and errors would be weeded out very quickly at runtime instead of surprising someone in production by accident.

Reply:
@alexcrichton Ah, I see. Having to add that parameter to many different types does indeed make this much less appealing.

One other alternative: could we consider a wasmtime feature flag for this, enabling/disabling async support, and then if it's enabled, we automatically use it for everything?

A concrete use case in my mind that's motivating the attempt to make this more than just a runtime flag: suppose you're writing a library crate that works with wasmtime and you provide functions to be called with wasmtime types, but you're not the top-level crate that's instantiating wasmtime. For instance, you want to write a library crate to expose some interesting functions, or to manage some objects. Conversely, the top-level crate may not *know* if it needs to call `new` or `new_async`, because it doesn't need any async itself but its dependencies might. If your crate requires async support, it seems unfortunate if that only lives in your crate's *documentation*, and you have to tell people "make sure to use `new_async` or you'll get runtime panics". I'd really love for the crate to be able to express, whether through the type system or through a cargo feature, "I need async and won't work without it".

If that's not possible, I understand. But if there's any way we can let crates surface "I need an async runtime" programmatically and ensure that they get one, rather than just adding it to their README, I think that'd be a usability improvement.

Reply:
I definitely agree that the scenario you're outlining is a downside. Unfortunately though I don't know of a fix other than two crates, `wasmtime` and `wasmtime-async`. Switching between async/not-async isn't really a good fit for Cargo features which are additive (but this is more mutually exclusive).

The only real way for embedders to support this in libraries would be to either pick one or to also support both async and sync APIs.

Put another way, I think a runtime flag is a local optimum for "don't duplicate the API surface area" as a constraint. Otherwise there are possibilities like different types or different crates I think, although I'm not sure which is best myself.